proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+</packageNamespaces>
+```
+
+Define the `<packageSource>` element within the `<packageNamespaces>` parent with the name of a valid package source:
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <packageSource name="nuget.org">
+
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+Next, add `<namespace>` elements under the `<packageSource>` with an `id` to specify the namespace. Namespaces can be defined as their full namepsace or using a wildcard(*) to match the glob pattern of 0 or more package ID(s):
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+If a user would like to support namespaces on different package sources, they can repeat this process for each source.
+A user may also pin a specific package id instead of the complete namespace.
+Specific ids take precedence over the namespaces.
+
+**Example:**
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+
+    <!-- Add a namespace for Contoso.* under the contoso source. -->
+    <packageSource name="contoso">
+        <namespace id="Contoso.*" />
+        <!-- Add a specific id to be download from  the contoso source. -->
+        <namespace id="Special.Package.With.An.Imperfect.Id" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+By default, namespaces will allow a package to match multiple namespaces and download with no precedence. In some cases, one might want to avoid ambiguous cases. That can be done by adding a `strict` flag.
+
+```xml
+<config>
+    <add key="namespaceMode" value="strict" />
+</config>
+```
+
+The strict flag provides a source pinning behavior by nature in which one namespace is allowed per package source. Entering a strict mode will provide users an error experience when there are namespace conflicts to which a user will be able to resolve by defining one namespace per package source configuration.
```

Comment:
Or rather, one package source is allowed per namespace.

Reply:
That's a great way to summarize it.

Reply:
I wanted to point out the error in "one namespace is allowed per package source", rather than summarize.

Reply:
Thanks! We'll fix it.

Reply:
This is still very opaque to me.  Does strict mode enforce that each `<packageSource>` can only contain a single `<namespace>`?  Does it mean that I can only include `<namespace id="Nuget.*" />` inside a single `<packageSource>`? Or is it saying that namespaced packages cannot alias to multiple sources, e.g. the following would cause errors for `Nuget.Internal.Package` because it matches namespaces in multiple different sources?
```xml
<packageSource name="nuget.org">
    <namespace id="Nuget.*" />
</packageSource>
<packageSource name="InternalFeed">
    <namespace id="Nuget.Internal.*" />
</packageSource>
```

Reply:
Let me update this shortly to help clarify. It means that only one unique package ID can be defined per package source. 

In your conflicting scenario, there is both `Nuget.*` & `Nuget.Internal.*`. However since you have `Nuget.Internal.*`, that takes precedence for the specific `Internal.*` scope & would therefore be resolved on the internal feed rather than the globbed nuget.org source. Everything outside the `Internal.*` scope would be resolved on nuget.org.

Reply:
I posted my comments above https://github.com/NuGet/Home/pull/10660/files#r597213005, but I don't think this will be easy to turn on if it's one package ID per package source. Basically any kind of organization that does pre-release builds on one set of feeds and releases on another will always have id overlap. The same packages will appear on both feeds and it's valid to use them on both feeds. This gets especially important in any kind of scenario where pre-release packages are different than released ones. For instance, the .NET 6 packages built day to day are unsigned, but those we release are signed, pushed to nuget.org, then deleted from our feeds so that nuget.org is the source of truth. This gets even more likely if you had more than one internal feed with the same packages. Perhaps legacy packages live on one, while new packages are pushed to another, etc. etc.

Could that allow for dependency confusion? Possibly, or at least there could be a restore that pulls from location X one time and location Y another. But when creating the namespaces config, the user is making a trust decision about namespaces on specific feeds. To me, `strict` should be a way of saying that a trust decision has been made about all referenced package IDs on all input sources.

If single namespace->single feed is still an important pivot, could we add an additional mode? Example
- `fullySpecified` - Any package id must be in one or more matching namespace declarations
- `singleSource` - No package id may match more than one feed (based on precedence rules)